include/crpropa/Grid.h

```diff
@@ -340,217 +318,183 @@ T interpolate(const Vector3d &position) {
-                                  CubicInterpolateScalar(periodicGet(ix,  iY,  iz-1),periodicGet(ix,  iY,  iz),periodicGet(ix,  iY,  iZ),periodicGet(ix,  iY,  iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(ix,  iy+2,iz-1),periodicGet(ix,  iy+2,iz),periodicGet(ix,  iy+2,iZ),periodicGet(ix,  iy+2,iz+2),posz),
-                                  posy),
-                 CubicInterpolateScalar(CubicInterpolateScalar(periodicGet(iX,  iy-1,iz-1),periodicGet(iX,  iy-1,iz),periodicGet(iX,  iy-1,iZ),periodicGet(iX,  iy-1,iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(iX,  iy,  iz-1),periodicGet(iX,  iy,  iz),periodicGet(iX,  iy,  iZ),periodicGet(iX,  iy,  iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(iX,  iY,  iz-1),periodicGet(iX,  iY,  iz),periodicGet(iX,  iY,  iZ),periodicGet(iX,  iY,  iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(iX,  iy+2,iz-1),periodicGet(iX,  iy+2,iz),periodicGet(iX,  iy+2,iZ),periodicGet(iX,  iy+2,iz+2),posz),
-                                  posy),
-                 CubicInterpolateScalar(CubicInterpolateScalar(periodicGet(ix+2,iy-1,iz-1),periodicGet(ix+2,iy-1,iz),periodicGet(ix+2,iy-1,iZ),periodicGet(ix+2,iy-1,iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(ix+2,iy,  iz-1),periodicGet(ix+2,iy,  iz),periodicGet(ix+2,iy,  iZ),periodicGet(ix+2,iy,  iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(ix+2,iY,  iz-1),periodicGet(ix+2,iY,  iz),periodicGet(ix+2,iY,  iZ),periodicGet(ix+2,iY,  iz+2),posz),
-                                  CubicInterpolateScalar(periodicGet(ix+2,iy+2,iz-1),periodicGet(ix+2,iy+2,iz),periodicGet(ix+2,iy+2,iZ),periodicGet(ix+2,iy+2,iz+2),posz),
-                                  posy),
-                 posx);
-	
+
+    double fX = r.x - floor(r.x);
+		double fY = r.y - floor(r.y);
+		double fZ = r.z - floor(r.z);
+
+    int nrCubicInterpolations = 4;
+		double interpolateVaryX[nrCubicInterpolations];
+		double interpolateVaryY[nrCubicInterpolations];
+		double interpolateVaryZ[nrCubicInterpolations];
+		/** Perform 1D interpolations while iterating in each for loop over the index of another direction */
+		for (int iLoopX = -1; iLoopX < nrCubicInterpolations-1; iLoopX++) {
+      for (int iLoopY = -1; iLoopY < nrCubicInterpolations-1; iLoopY++) {
+        for (int iLoopZ = -1; iLoopZ < nrCubicInterpolations-1; iLoopZ++) {
+          interpolateVaryZ[iLoopZ+1] = periodicGet(iX0+iLoopX, iY0+iLoopY, iZ0+iLoopZ);
+        }
+        interpolateVaryY[iLoopY+1] = CubicInterpolateScalar(interpolateVaryZ[0], interpolateVaryZ[1], interpolateVaryZ[2], interpolateVaryZ[3], fZ);
+      }
+      interpolateVaryX[iLoopX+1] = CubicInterpolateScalar(interpolateVaryY[0], interpolateVaryY[1], interpolateVaryY[2], interpolateVaryY[3], fY);
+    }
```

Comment:
I just ran the test comparing this to the previous version (only the non-SIMD code, maybe we should do the same test for the SIMD as well but I'll see when I can get to that), and it looks good. However, I found a bug in setInterpolationType (see above)